async-2.6: Move async into experimental

pwinston opened the pull request:
# Description

This is a **wedge PR** for async rendering. It wedges in the **async rendering** functionality by making a few strategic changes mostly related to the `Image` class, while moving most async related functionality into a new `experimental` namespace. The `experimental` code is not even imported unless `NAPARI_ASYNC` is defined.

The intent is to protect regular users from any problems with the async code. Unless you opt-in to async you are not running that code at all. After this wedge PR, the hope is the PR's `async-3` through `async-12` are easier to review since it will be clear which changes apply only to experimental code.

## Changes:

1. Move `chunk` module into `napari.components.experimental`.
2. Move `QtChunkReceiver` class into `napari._qt.experimental`.
3. New async `ChunkedImageLoader` and `ChunkedSliceData` classes in `napari.layers.image.experimental`.
4. New legacy `ImageLoader` and `ImageSliceData` classes in `napari.layers.image` (for non-async).

## When NAPARI_ASYNC is defined

1. `viewer.Viewer.close()` calls into `chunk_loader` to cleanup.
2. `QtLayerList.__init__()` creates a `QtChunkReceiver`.
3. `Image` gets one extra method: `Image.on_chunk_loaded()`.
4. `Image` uses `ChunkedSliceData` instead of `ImageSliceData`.
5. `ImageSlice` uses `ChunkedImageLoader` instead of `ImageLoader`.

**Notes:**
1. Only `QtChunkReceiver` calls into the new `Image.on_chunk_loaded()` method, the only purpose of `on_chunk_loaded()` is to receive chunks from the `ChunkLoader` via the `QtChunkReceiver`.
2. `ChunkedSliceData` derives from `ImageSliceData` to add two async related methods: `load_chunks` and `from_request`. Both are only used when a `ChunkRequest` was asynchronously loaded.
3. `ChunkedImageLoader` derives from `Imageloader` and overrides the `load()` and `match()` methods for async.

We use `Chunked` prefix here instead of `Async` because `NAPARI_ASYNC` really means "load with the `ChunkLoader`" but those loads might be synchronous or asynchronous. We do synchronous loads for `ndarray` and for fast loading arrays of any type. So `ChunkedImageLoader` is a better than name `AsyncImageLoader` since the loads will not always be async.

## Type of change
- [x] Refactoring, moving code into experimental, no user visible changes.

pwinston committed:
Fix async after _new_empty_slice() change

We added _new_empty_slice() in master to fix labels. But
this actually broke async because in async we were calling
refresh when the chunks was loaded which emptied the slice.

For now we just called events.set_data() and _update_thumbnail()
manually. Will need more thought.

pwinston committed:
Prep for most to experimental

Create ImageSliceData, SyncImageLoader and AsyncImageLoader
So that we can have a clear line between master/sync and experimental/async

pwinston committed:
Fix multiscale typo

pwinston committed:
Move perf into components.experimental.perf

pwinston committed:
Fix method name

pwinston committed:
Fix use_async test!

pwinston committed:
Blank __init__.py needed for pytest

pwinston committed:
Move QtChunkReceiver into experimental

pwinston committed:
Slim down comments and normalize them.

pwinston committed:
Fully guard it with NAPARI_ASYNC

pwinston committed:
Only create global if NAPARI_ASYNC

pwinston committed:
Disable test temporarily

pwinston committed:
Add _use_async

pwinston committed:
Slim down comment

pwinston committed:
Move experimental import down

pwinston committed:
Add experimental AsyncImageSliceData

pwinston committed:
Do not import experimental at all unless async

pwinston committed:
if statement not needed

pwinston committed:
Guard experimental code against import

pwinston committed:
Derive from ImageLoader

Not really needed, but to be consistent with
AsyncImageSliceData and ImageSliceData.

pwinston committed:
Comments, rename chunk_request -> request

pwinston committed:
Check for pytest

pytest imports everything,

pwinston committed:
Detect with sys.modules?

pwinston committed:
Remove tests for now

These tests made sure we don't import async code
unless NAPARI_ASYNC is defined. But they are failing
in CI (pytest?). It must import everything regradless
of whether it is referenced.

Might be a way around this but remove it for now.

pwinston committed:
More doc string

pwinston committed:
Rename some Async classes to Chunked

Although the feature toggle is NAPARI_ASYNC the
ChunkLoader does both sync and async loading.

So better to call these Chunked not Async.

pwinston committed:
Docs and return bool instead of data

Returning the data was just not necessary.

pwinston committed:
Fix name 'Chunked'

pwinston committed:
Doc string and fix return value

pwinston committed:
Blank since others are blang

pwinston committed:
Use forward reference for return type

sofroniewn commented on napari/layers/image/image.py:
```diff
@@ -645,18 +635,19 @@ def on_chunk_loaded(
         # didn't match the rgb flag that the user specified (or we inferred?).
-        if self.rgb != guess_rgb(request.image.shape):
+        # TODO_ASYNC: do something sooner so this check is not required?
+        if self.rgb != data.rgb:
```
I know this isn't a change in this PR but, it was already added in an earlier async PR but I want to flag that this still causes a bug in the synchronous code that didn't exist before the async work started (i.e. roll back to 0.3.6 and the following works fine). To reproduce the bug just run

```python
    viewer = napari.view_image(data.astronaut(), rgb=False, ndisplay=3)
```

When it was added to async I think the justification was something like it may have prevented an error, but given that this introduces a bug where there wasn't one before I think we should probably just drop it.

pwinston replied:
Okay I removed the rgb test entirely. When that error comes up again we can try to find a better way to flag it and help the user figure out what went wrong.

sofroniewn replied:
sounds good, thanks!

sofroniewn commented on napari/layers/image/_image_slice_data.py:
```diff
@@ -0,0 +1,64 @@
+
+    @property
+    def rgb(self) -> bool:
+        return guess_rgb(self.image.shape)
```
See comment below about dropping

pwinston replied:
Removed

sofroniewn reviewed (approved):
Ok, this looks good to me and I'm approving. I've flagged one more bug around the `rgb` guessing, that I think we should fix before merge, but that is just dropping a line of code. I've run this PR through my battery of interactive examples, including multiscale and painting and not found any other problems.

Overall I also like the better separation between experimental and non-experimental code, thanks for that.

I will flag though the addition of two objects `ImageSliceData` and `ImageLoader` which now adds some indirection to the sync code to make the parallelism between async and sync easier. I think this is ok as the objects are very small and easy to understand, and the data flow is clear to me, but at some point when we have everything working we might want to look into reducing the number of objects if it makes things simpler again.

Along that vein I will also flag that I think a lot of the additional complexity comes from the "thumbnail" and the need to carry around the additional "thumbnail source" which is only relevant for multiscale data (where we use one of the low resolution levels for the thumbnail). Maybe when we get to the octree we can think if there are ways to reduce some of the complexity. In general we probably want to revisit thumbnail generation for all the layers too.

I see @pwinston you mentioned you might want to look into one or two other things tomorrow before I merge, so I'll wait for you to give the green light, and then it will be good to go!

pwinston committed:
Remove rgb test for now

This breaks something with sync. This test was added
to avoid getting some mysterious error when your
rgb flag was wrong. If that error comes up again
we can try to figure out a way to flag it earlier
before the load finishes.
